### PR TITLE
[2520] feat(model): Refactor EMA update

### LINKED ISSUE #2520: Optimize EMA weight updates to reduce training step overhead
### Describe the task. It can be a feature, documentation, etc.

Recent profiling traces reviewed with Nvidia show that the EMA model update is taking up roughly 10% of our step time.

The EMA model is updated synchronously at every step here:
https://github.com/ecmwf/WeatherGenerator/blob/1468a5016314d00734fecc7b6fb1f2696fc29862/src/weathergen/train/trainer.py#L529-L530

The EMA model weights update does not affect the weights update of the original model, which is more like a 'validation' step. Given that we already store all weights in our checkpoints, we should evaluate if this 10% overhead is necessary in most configurations, for example:
https://github.com/ecmwf/WeatherGenerator/blob/1468a5016314d00734fecc7b6fb1f2696fc29862/config/config_operan_georing_avhrr_forecasting_lowres.yml#L204-L205

A few ideas about optimization:
1. EMA weights can be generated in the post-processing or updated asynchoronously.
2. The interval of EMA weights update can be every 10 steps or larger. The hyperparameters should be tuned accordingly.

Related PR:
https://github.com/ecmwf/WeatherGenerator/pull/1005

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

https://gitlab.jsc.fz-juelich.de/hedgedoc/Lxcjjeu5TgeeNdlRbMgv4A?both

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [x] performance

### PR DESCRIPTION
**Pull request by florianscheidl**
## Description

- Compute EMA param updates in place with a fused torch implementation
- Move the EMA source/parameter pair check outside the update function to reset.

We profiled the old and the new implementation. On the config_operan_georing_avhrr_forecasting_lowres, an update took:
- previously, ~100ms (oan0f4qo on Santis): Screenshot 2026-06-24 at 15 05 51
- afterwards, ~10ms (h3ofhgh5 on Santis): Screenshot 2026-06-24 at 15 06 07

## Issue Number

Fixes #2520.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Comment by florianscheidl** [comment]
@ekouts

**Review by sophie-xhonneux — APPROVED** [review]
I presume you tested this with DDP and FSDP with any kind of JEPA config? If yes, feel free to merge. Conceptually this PR seems great! Please also open it against develop-ssl.

**Comment by florianscheidl** [comment]
> I presume you tested this with DDP and FSDP with any kind of JEPA config? If yes, feel free to merge. Conceptually this PR seems great! Please also open it against develop-ssl.

Hey, thx for the quick review. We've tested it with config_operan_georing_avhrr_forecasting_lowres.yml (with fsdp). If you like, we can run with config_jepa.

**Review by clessig — CHANGES_REQUESTED** [review]
Do we have some loss curves (even a few epochs is fine) with the PR.

**Comment by florianscheidl** [comment]
> Do we have some loss curves (even a few epochs is fine) with the PR.

Not yet. I can do a run before and after and plot some.

**Comment by clessig** [comment]
> > Do we have some loss curves (even a few epochs is fine) with the PR.
> 
> Not yet. I can do a run before and after and plot some.

Thanks, that would be great!

**Comment by florianscheidl** [comment]
I created a PR to the develop-ssl branch here: https://github.com/ecmwf/WeatherGenerator/pull/2580, plotted training losses (since we trained a jepa config).
Will do the same for this branch, but plot validation curves using a forecasting config.

**Comment by florianscheidl** [comment]
> I created a PR to the develop-ssl branch here: #2580, plotted training losses (since we trained a jepa config). Will do the same for this branch, but plot validation curves using a forecasting config.

Loss curves (64 samples per mini-epoch):
Screenshot 2026-07-02 at 16 52 55

Looks a bit peculiar that "before" is faster here. Might hardly make a difference for smaller configs or even slow it down a bit. In the larger configs, we saw a slight performance improvement.
